File: src/test/java/com/addressbook/test/LoginLogoutWithRolesTest.java
package com.addressbook.test;

import org.junit.Assert;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class LoginLogoutWithRolesTest {

    @Test
    public void stage1_loginLogoutWithRoles() throws InterruptedException {
        // Set Chrome driver location
        System.setProperty("webdriver.chrome.driver", LoginLogoutWithRolesTest.class.getClassLoader().getResource("chromedriver").getPath());
        // Initialize Selenium driver
        WebDriver driver = new ChromeDriver();
        // Initialize wait driver
        WebDriverWait webDriverWait = new WebDriverWait(driver, 20);
        // Open login page
        driver.get("http://localhost:8080");
        // Wait until page is loaded
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"application\"]")));
        // Locate login input field
        WebElement loginInput = driver.findElement(By.xpath("//*[@id=\"login\"]"));
        // Locate password input field
        WebElement passwordInput = driver.findElement(By.xpath("//*[@id=\"password\"]"));
        // Locate login button
        WebElement loginButton = driver.findElement(By.xpath("//*[@id=\"application\"]/div/div/div[2]/div[2]/button"));
        // Enter login for ordinary user
        loginInput.sendKeys("user");
        // Enter password
        passwordInput.sendKeys("userPass");
        // Click login
        loginButton.click();
        // Locate user info button on navigation bar
        WebElement userName = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"application\"]/div/div/nav/div/div/ul[2]/div/button")));
        // Check if user name is right
        Assert.assertEquals(userName.getText(), "user");
        // Click user info button to check roles
        userName.click();
        // Locate first role
        WebElement userRole = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div/div[2]/div/div/div[2]/span")));
        /// This user should have only one role called USER, check that
        Assert.assertEquals(userRole.getText(), "USER");
        // Close modal with user roles
        driver.findElement(By.xpath("/html/body/div[2]/div/div[2]/div/div/div[1]/button")).click();
        // Wait for modal window to close
        Thread.sleep(300);
        // Click logout button
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"application\"]/div/div/nav/div/div/ul[2]/button"))).click();
        // Login as admin
        loginInput = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"login\"]")));
        loginInput.sendKeys("admin");
        passwordInput = driver.findElement(By.xpath("//*[@id=\"password\"]"));
        passwordInput.sendKeys("adminPass");
        loginButton = driver.findElement(By.xpath("//*[@id=\"application\"]/div/div/div[2]/div[2]/button"));
        loginButton.click();
        userName = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"application\"]/div/div/nav/div/div/ul[2]/div/button")));
        Assert.assertEquals(userName.getText(), "admin");
        userName.click();
        userRole = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div/div[2]/div/div/div[2]/span")));
        Assert.assertEquals(userRole.getText(), "USER");
        WebElement adminRole = driver.findElement(By.xpath("/html/body/div[2]/div/div[2]/div/div/div[2]/span[2]"));
        // Admin should also have second role called ADMIN, check that
        Assert.assertEquals(adminRole.getText(), "ADMIN");
        // Close modal with user roles
        driver.findElement(By.xpath("/html/body/div[2]/div/div[2]/div/div/div[1]/button")).click();
        // Wait for modal window to close
        Thread.sleep(300);
        // Click logout button
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"application\"]/div/div/nav/div/div/ul[2]/button"))).click();
        Thread.sleep(300);
        driver.quit();
    }
}

File: addressbook-main/src/test/kotlin/com/addressbook/test/LoginLogoutWithRolesTest.kt
package com.addressbook.test

import io.github.bonigarcia.wdm.WebDriverManager
import org.junit.Assert
import org.junit.Test
import org.openqa.selenium.By
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.chrome.ChromeOptions
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.WebDriverWait
import java.awt.Robot
import java.awt.event.KeyEvent
import java.time.Duration

class LoginLogoutWithRolesTest {

    @Test
    fun stage1_loginLogoutWithRoles() {
        WebDriverManager.chromedriver().setup()
        // Initialize Selenium driver
        val options = ChromeOptions()
        options.addArguments("--remote-allow-origins=*")
        val driver = ChromeDriver(options)
        // Initialize wait driver
        val webDriverWait = WebDriverWait(driver, Duration.ofSeconds(20))
        // Dismiss certificate choice dialog
        Thread {
            try {
                Thread.sleep(1_000)
                val robot = Robot()
                robot.keyPress(KeyEvent.VK_ESCAPE)
                robot.keyRelease(KeyEvent.VK_ESCAPE)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }.start()
        // Open login page
        driver.get("https://localhost:9000")
        // Wait until page is loaded
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"login\"]")))
        // Locate login input field
        var loginInput = driver.findElement(By.xpath("//*[@id=\"login\"]"))
        // Locate password input field
        var passwordInput = driver.findElement(By.xpath("//*[@id=\"password\"]"))
        // Locate login button
        var loginButton = driver.findElement(By.xpath("/html/body/div[2]/div[3]/div/div[2]/div/button"))
        // Enter login for ordinary user
        loginInput.sendKeys("user")
        // Enter password
        passwordInput.sendKeys("userPass")
        // Click login
        loginButton.click()
        // Locate user info button on navigation bar
        var userName = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div/div/div/div/header/div/div/button[1]")))
        // Check if username is right
        Assert.assertEquals("USER", userName.text)
        // Click user info button to check roles
        userName.click()
        // Locate first role
        var userRole = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div[3]/div/div[1]/div[1]/span")))
        /// This user should have only one role called USER, check that
        Assert.assertEquals("USER", userRole.text)
        // Close modal with user roles
        Thread {
            try {
                val robot = Robot()
                robot.keyPress(KeyEvent.VK_ESCAPE)
                robot.keyRelease(KeyEvent.VK_ESCAPE)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }.start()
        // Click logout button
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div/div/div/div/header/div/div/button[4]"))).click()
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div[3]/div/div[2]/button[2]"))).click()
        // Login as admin
        loginInput = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//*[@id=\"login\"]")))
        loginInput.sendKeys("admin")
        passwordInput = driver.findElement(By.xpath("//*[@id=\"password\"]"))
        passwordInput.sendKeys("adminPass")
        loginButton = driver.findElement(By.xpath("/html/body/div[2]/div[3]/div/div[2]/div/button"))
        loginButton.click()
        userName = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div/div/div/div/header/div/div/button[1]")))
        Assert.assertEquals("ADMIN", userName.text)
        userName.click()
        userRole = webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div[3]/div/div[1]/div[1]/span")))
        Assert.assertEquals("USER", userRole.text)
        val adminRole = driver.findElement(By.xpath("/html/body/div[2]/div[3]/div/div[1]/div[2]/span"))
        // Admin should also have second role called ADMIN, check that
        Assert.assertEquals("ADMIN", adminRole.text)
        // Close modal with user roles
        Thread {
            try {
                val robot = Robot()
                robot.keyPress(KeyEvent.VK_ESCAPE)
                robot.keyRelease(KeyEvent.VK_ESCAPE)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }.start()
        // Click logout button
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div/div/div/div/header/div/div/button[4]"))).click()
        webDriverWait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div[2]/div[3]/div/div[2]/button[2]"))).click()
        driver.quit()
    }
}
